src/main/kotlin/org/rust/ide/intentions/ImportNameIntention.kt

```diff
@@ -0,0 +1,232 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.ide.intentions
+
+import com.intellij.codeInsight.intention.HighPriorityAction
+import com.intellij.codeInsight.navigation.NavigationUtil
+import com.intellij.ide.util.DefaultPsiElementCellRenderer
+import com.intellij.openapi.command.WriteCommandAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.ui.popup.PopupStep
+import com.intellij.openapi.ui.popup.util.BaseListPopupStep
+import com.intellij.psi.PsiElement
+import com.intellij.ui.popup.list.ListPopupImpl
+import com.intellij.ui.popup.list.PopupListElementRenderer
+import com.intellij.util.ObjectUtils
+import com.intellij.util.ui.UIUtil
+import org.rust.cargo.icons.CargoIcons
+import org.rust.cargo.project.workspace.PackageOrigin
+import org.rust.cargo.util.AutoInjectedCrates
+import org.rust.ide.icons.RsIcons
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.stubs.index.RsNamedElementIndex
+import org.rust.lang.core.types.ty.TyPrimitive
+import org.rust.openapiext.Testmark
+import java.awt.BorderLayout
+import java.awt.Component
+import javax.swing.*
+
+class ImportNameIntention : RsElementBaseIntentionAction<ImportNameIntention.Context>(), HighPriorityAction {
+
+    override fun getText() = "Import"
+    override fun getFamilyName() = text
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val path = element.ancestorStrict<RsPath>() ?: return null
+        if (TyPrimitive.fromPath(path) != null || path.reference.resolve() != null) return null
+
+        val parent = path.path
+        val parentRes = parent?.reference?.resolve()
+        if (parent == null || parentRes is RsMod || parentRes is RsEnumItem) {
+            val basePath = getBasePath(path)
```

Comment:
I am a bit confused over the logic here. We can autoimport only the first component of the path. However, we can invoke intention on any part of the path

```
foo::bar::/*caret*/baz
^^^ unresolved
```

This sounds just right! What I can't understand is why do we try to resolve the first qulifier here? 


```
foo::bar::baz::quux::fr/*caret*/ob;
^               ^- why do we try to resolve this component?
| - this is what we want to auto import
```

Shouldn't the logic be like this instead:

```
base = getBasePath(path)
if (base.reference.resolve() != null) return
try to import base
```

Reply:
Yeah, it makes sense!